### PR TITLE
fix ArrayWidget so it stores numbers to formData when needed

### PR DESCRIPTION
**Pull request by mplis-jetsetter**

### CODE REVIEW

**Comment by mplis-jetsetter** [comment]
Attempts to fix #295

File: src/components/widgets/SelectWidget.js [hunk 1]
```diff
@@ -7,13 +7,23 @@ import { asNumber } from "../../utils";
...
  * always retrieved as strings.
  */
-function processValue(type, value) {
-  if (type === "boolean") {
-    return value === "true";
-  } else if (type === "number") {
-    return asNumber(value);
+function processValue(schema, value) {
+  if (schema.type === "array") {
+    let valueType = schema.items.type;
```
Comment (n1k0) [inline src/components/widgets/SelectWidget.js]: Let's use `const`

File: src/components/widgets/SelectWidget.js [hunk 2]
```diff
@@ -7,13 +7,23 @@ import { asNumber } from "../../utils";
...
+    if (valueType === "number" || valueType === "integer") {
+      return value.map(v => asNumber(v));
+    }
+    return value;
+  }
+  else {
```
Comment (n1k0) [inline src/components/widgets/SelectWidget.js]: Nit: unnecessary `else` clause as we've returned early.

File: test/ArrayField_test.js [hunk 3]
```diff
@@ -599,4 +599,31 @@ describe("ArrayField", () => {
...
+      Simulate.change(node.querySelector(".field select"), {
+        target: {options: [
+          {selected: true, value: "1"},
+          {selected: true, value: "2"},
+          {selected: false, value: "3"},
+        ]}
+      });
+
+      expect(comp.state.formData).eql([1, 2]);
+
```
Comment (n1k0) [inline test/ArrayField_test.js]: Nit: please remove this blank line

**Comment by n1k0** [comment]
Great patch, thank you! r+ with comments addressed

File: src/components/widgets/SelectWidget.js [hunk 4]
```diff
@@ -9,21 +9,19 @@ import { asNumber } from "../../utils";
...
-      return asNumber(value);
-    }
-    return value;
+  const valueType = schema.type;
+  if (valueType === "boolean") {
+    return value === "true";
+  } else if (valueType === "number") {
+    return asNumber(value);
   }
+  return value;
```
Comment (n1k0) [inline src/components/widgets/SelectWidget.js]: Actually I suspect we could rewrite the whole function in a more compact fashion, like:

```js
function processValue({type, items}, value) {
  if (type === "array" && items && ["number", "integer"].includes(items.type)) {
    return value.map(asNumber);
  } else if (type === "boolean") {
    return value === "true";
  } else if (type === "number") {
    return asNumber(value);
  }
  return value;
}
```

**Comment by n1k0** [comment]
Sorry for the late bit of review ;)

**Comment by mplis-jetsetter** [comment]
No worries, yours is much better.

Side note: is there a way to run a specific test? The full test suite takes several minutes to run and sometimes I just want to check if a change I made causes a specific test to fail.

**Comment by n1k0** [comment]
> is there a way to run a specific test? 

Yeah, you can use `it.only()` to only run a single test. Combined with `npm run tdd` which will rerun the test everytime the source file changes, that's really useful to work on some implementation.

> The full test suite takes several minutes to run

Wow, several minutes? Doesn't sound good :/

**Comment by n1k0** [comment]
I'm landing this, great work, thanks!

**Comment by n1k0** [comment]
Released in [v0.38.1](https://github.com/mozilla-services/react-jsonschema-form/releases/tag/v0.38.1).

**Pull request merged**